olp-cpp-sdk-dataservice-read/src/repositories/DataRepository.cpp

```diff
@@ -49,6 +50,74 @@ constexpr auto kBlobService = "blob";
 constexpr auto kVolatileBlobService = "volatile-blob";
 }  // namespace
 
+DataResponse DataRepository::GetVersionedDataTileQuadTree(
+    const client::HRN& catalog, const std::string& layer_id,
+    TileRequest request, int64_t version, client::CancellationContext context,
+    client::OlpClientSettings settings) {
+  auto tile = request.GetTileKey().ToHereTile();
+  std::string requested_tile_data_handle;
+
+  auto cached_partitions = PartitionsRepository::GetTileFromCache(
+      catalog, layer_id, request, version, settings);
+  if (cached_partitions.GetPartitions().size() > 0) {
+    OLP_SDK_LOG_INFO_F(kLogTag, "cache data '%s' found!",
+                       request.CreateKey(layer_id).c_str());
+
+    for (const auto& partition : cached_partitions.GetPartitions()) {
+      // find data handle for requested tile
+      if (partition.GetPartition().compare(tile) == 0) {
+        requested_tile_data_handle = partition.GetDataHandle();
+        OLP_SDK_LOG_INFO_F(kLogTag, "Requested tile data handle: %s.",
+                           requested_tile_data_handle.c_str());
+        break;
+      }
+    }
+  } else if (request.GetFetchOption() == CacheOnly) {
+    OLP_SDK_LOG_INFO_F(kLogTag, "cache tile '%s' not found!",
+                       request.CreateKey(layer_id).c_str());
+    return ApiError(ErrorCode::NotFound,
+                    "Cache only resource not found in cache (data).");
+  } else {
+    auto parent_tile = request;
+    parent_tile.WithTileKey(request.GetTileKey().Parent());
+    auto cached_partitions = PartitionsRepository::GetTileFromCache(
+        catalog, layer_id, parent_tile, version, settings);
+    if (cached_partitions.GetPartitions().size() != 0) {
+      DataRequest data_request =
+          DataRequest()
+              .WithPartitionId(request.GetTileKey().ToHereTile())
+              .WithVersion(version)
+              .WithFetchOption(request.GetFetchOption());
+      // get the requested tile data, if parent tile was found in cache
+      return repository::DataRepository::GetVersionedData(
+          catalog, layer_id, data_request, context, settings);
+    }
+
+    auto response = PartitionsRepository::QueryPartitionsAndGetDataHandle(
+        catalog, layer_id, request, version, context, settings,
+        requested_tile_data_handle);
+    if (!response.IsSuccessful()) {
+      return response.GetError();
+    }
+  }
```

Comment:
All this should not be here, it should be implementation details of the `PartitionsRepository::GetVersionedTile()`.

Reply:
Please check the other methods in this class, partition repo works with the metadata only.

Reply:
This is an implementation detail of PartitionsRepo which is easily fixed, considering that the query REST API wrapper is static and can be used everywhere.

Reply:
Moved to PartitionRepository